shopinvader_product/models/product_product.py

```diff
@@ -0,0 +1,211 @@
+        short_name = ", ".join(attributes.mapped("name"))
+        full_name = self.display_name
+        if short_name:
+            full_name += " (%s)" % short_name
+        return full_name, short_name
+
+    def _compute_names(self):
+        for record in self:
+            (
+                record.full_name,
+                record.short_name,
+            ) = record._prepare_variant_name_and_short_name()
+
+    @api.model
+    def _get_shopinvader_product_variants(self, product_ids):
+        # Use sudo to bypass permissions (we don't care)
+        return self.sudo().search(
+            [("product_tmpl_id", "in", product_ids)], order="product_tmpl_id"
+        )
+
+    def _compute_main_product(self):
+        # Respect same order.
+        order_by = [x.strip() for x in self.env["product.product"]._order.split(",")]
+        fields_to_read = ["product_tmpl_id"] + [f.split(" ")[0] for f in order_by]
+        product_ids = self.mapped("product_tmpl_id").ids
+        _variants = self._get_shopinvader_product_variants(product_ids)
+        # Use `load=False` to not load template name
+        variants = _variants.read(fields_to_read, load=False)
+        var_by_product = groupby(variants, lambda x: x["product_tmpl_id"])
+
+        def pick_1st_variant(variants):
+            def get_value(record, key):
+                if record[key] is False and self._fields[key].type in ("char", "text"):
+                    return ""
+                else:
+                    return record[key]
+
+            for order_key in reversed(order_by):
+                order_key_split = order_key.split(" ")
+                reverse = len(order_key_split) > 1 and order_key_split[1] == "desc"
+                variants.sort(
+                    key=lambda var: get_value(var, order_key_split[0]),
+                    reverse=reverse,
+                )
+            return variants[0].get("id") if variants else None
+
+        main_by_product = {
+            product: pick_1st_variant(list(variants))
+            for product, variants in var_by_product
+        }
+        for record in self:
+            record.main = main_by_product.get(record.product_tmpl_id.id) == record.id
+
+    @api.depends_context("pricelist", "default_role", "fposition", "company", "date")
+    def _compute_price(self):
+        for record in self:
+            record.price = record._get_all_price()
+
+    def _get_all_price(self):
+        self.ensure_one()
+        res = {}
+        pricelist = self._context.get("pricelist")
+        default_role = self._context.get("default_role", "default")
+        if pricelist:
+            fposition = self._context.get("fposition")
+            company = self._context.get("company")
+            date = self._context.get("date")
+            res[default_role] = self._get_price(
+                pricelist=pricelist, fposition=fposition, company=company, date=date
+            )
+        return res
+
+    def _get_price(
```

Comment:
are you sure it's still needed ? 

there was some changes between v14 and v16 in product regarding the price calculation